controllers/s3utils.go

```diff
@@ -0,0 +1,404 @@
+// bucket := "subname-namespace" // should be all lowercase
+// if err := s3Conn.createBucket(bucket); err != nil {
+// 	fmt.Println(errorswrapper.Wrapf(err, "Unable to create a bucket: %s", bucket))
+// 	return
+// }
+
+// *** Upload objects, optionally using a key prefix to easily find the objects later ***
+// for i := 1; i < 10; i++ {
+// 	pvKey := fmt.Sprintf("PersistentVolumes/pv%v", i)
+// 	uploadPV := corev1.PersistentVolume{}
+// 	uploadPV.Name = pvKey
+// 	uploadPV.Spec.StorageClassName = "gold"
+// 	uploadPV.Spec.PersistentVolumeReclaimPolicy = corev1.PersistentVolumeReclaimRetain
+// 	if err := s3Conn.uploadObject(bucket, pvKey, uploadPV); err != nil {
+// 		fmt.Println(errorswrapper.Wrapf(err, "Unable to upload: %s:%s", bucket, pvKey))
+// 		return
+// 	}
+// }
+
+// *** Find objects in the bucket, optionally supplying a key prefix
+// keyPrefix := "v1.PersistentVolumes/"
+// if list, err := s3Conn.listKeys(bucket, keyPrefix); err != nil {
+// 	fmt.Println(errorswrapper.Wrapf(err, "Unable to download: %s:%s", bucket, keyPrefix))
+// 	return
+// } else {
+// 	for _, key := range list {
+// 		fmt.Printf("%v ", key)
+// 	}
+// }
+
+// *** Download from the given bucket an object with the given key
+// keyPrefix := "v1.PersistentVolumes/"
+// key := keyPrefix + "pv2"
+// var downloadPV corev1.PersistentVolume
+// if err := s3Conn.downloadObject(bucket, key, &downloadPV); err != nil {
+// 	fmt.Println(errorswrapper.Wrapf(err, "Unable to download: %s:%s", bucket, key))
+// 	return
+// }
+// }
+
+type s3Connection struct {
+	session    *session.Session
+	client     *s3.S3
+	uploader   *s3manager.Uploader
+	downloader *s3manager.Downloader
+	endpoint   string
+	callerTag  string
+}
+
+var objectStoreConnections = map[string]*s3Connection{}
+
+// Return a connection to the given S3 object store endpoint using
+// the given s3 secret
+// - Return error if endpoint or secret is not configured, or if
+//   session creation fails
+func connectToS3Endpoint(ctx context.Context, r client.Reader,
+	s3Endpoint string, s3SecretName types.NamespacedName,
+	callerTag string) (*s3Connection, error) {
+	if s3Endpoint == "" {
+		return nil, fmt.Errorf("s3 endpoint has not been configured; tag:%v",
+			callerTag)
+	}
+
+	// Use cached connection, if one exists
+	if s3Conn, ok := objectStoreConnections[s3Endpoint]; ok {
+		return s3Conn, nil
+	}
+
+	accessID, secretAccessKey, err := getS3Secret(ctx, r, s3SecretName)
+	if err != nil {
+		return nil, errorswrapper.Wrapf(err,
+			"failed to get secret %v; tag %v",
+			s3SecretName, callerTag)
+	}
+
+	// Create an S3 client session
+	s3Session, err := session.NewSession(&aws.Config{
+		Credentials: credentials.NewStaticCredentials(string(accessID),
+			string(secretAccessKey), ""),
+		Endpoint:         aws.String(s3Endpoint),
+		Region:           aws.String("us-east-1"),
+		DisableSSL:       aws.Bool(true),
+		S3ForcePathStyle: aws.Bool(true),
+	})
+	if s3Session == nil || err != nil {
```

Comment:
Isn't the pattern is to test for `err` only?  In other words, if `NewSession` returns error, then s3Session is guaranteed to be nil otherwise, it is guaranteed to be a valid session.

Reply:
Right; I will change it to only test for err.